crates/bevy_ecs/src/schedule/system_graph.rs

```diff
@@ -0,0 +1,264 @@
+/// let graph = SystemGraph::new();
+///
+/// let start_a = graph.root(sys_a.system());
+/// let start_b = graph.root(sys_b.system());
+/// let start_c = graph.root(sys_c.system());
+///
+/// vec![start_a, start_b, start_c]
+///     .join_into(sys_d.system())
+///     .then(sys_e.system());
+///
+/// // Convert into a SystemSet
+/// let system_set: SystemSet = graph.into();
+/// ```
+#[derive(Clone, Default)]
+pub struct SystemGraph(Arc<Mutex<HashMap<NodeId, SystemDescriptor>>>);
+
+impl SystemGraph {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn root(&self, system: impl Into<SystemDescriptor>) -> SystemGraphNode {
+        self.create_node(system.into())
+    }
+
+    pub fn is_same_graph(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.0, &other.0)
+    }
+
+    fn create_node(&self, mut system: SystemDescriptor) -> SystemGraphNode {
+        let id = NodeId(NEXT_NODE_ID.fetch_add(1, Ordering::Relaxed));
+        match &mut system {
+            SystemDescriptor::Parallel(descriptor) => descriptor.labels.push(id.dyn_clone()),
+            SystemDescriptor::Exclusive(descriptor) => descriptor.labels.push(id.dyn_clone()),
+        }
+        self.0.lock().unwrap().insert(id, system);
+        SystemGraphNode {
+            id,
+            graph: self.clone(),
+        }
+    }
+
+    fn add_dependency(&self, src: NodeId, dst: NodeId) {
+        if let Some(system) = self.0.lock().unwrap().get_mut(&dst) {
+            match system {
+                SystemDescriptor::Parallel(descriptor) => descriptor.after.push(src.dyn_clone()),
+                SystemDescriptor::Exclusive(descriptor) => descriptor.after.push(src.dyn_clone()),
+            }
+        } else {
+            panic!(
+                "Attempted to add dependency for {:?}, which doesn't exist.",
+                dst
+            );
+        }
+    }
+}
+
+/// A draining conversion from [SystemGraph] to [SystemSet]. All other clones of the same graph
+/// will be empty afterwards.
+///
+/// [SystemGraph]: crate::schedule::SystemSet
+/// [SystemSet]: crate::schedule::SystemSet
+impl From<SystemGraph> for SystemSet {
+    fn from(graph: SystemGraph) -> Self {
+        let mut system_set = SystemSet::new();
+        for (_, system) in graph.0.lock().unwrap().drain() {
+            system_set = system_set.with_system(system);
+        }
+        system_set
+    }
+}
+
+#[derive(Clone)]
+pub struct SystemGraphNode {
+    id: NodeId,
+    graph: SystemGraph,
+}
+
+impl SystemGraphNode {
+    pub fn then(&self, next: impl Into<SystemDescriptor>) -> SystemGraphNode {
+        let node = self.graph.create_node(next.into());
+        self.add_dependent(node.id);
+        node
+    }
+
+    fn add_dependent(&self, dst: NodeId) {
+        self.graph.add_dependency(self.id, dst);
+    }
+}
+
+pub trait SystemGraphJoinExt: Sized + IntoIterator<Item = SystemGraphNode> {
+    fn join_into(self, next: impl Into<SystemDescriptor>) -> SystemGraphNode {
```

Comment:
`join_into` is not very descriptive of what this is doing imo. 
Perhaps
`join_into_node`
`coalesce_system_nodes`
`merge_nodes`

Reply:
I used join because that's what the dataflow graph APIs used. It's close in both name and function to APIs like futures::join. 

Merge implies the existing nodes are consumed to make the the new one. Whereas it's actually just linking the new system to all of its dependencies. Not sure about that one.

Reply:
perhaps then just `join`?

Reply:
Sounds good to me. Matches well with `fork`.